services/web/server/tests/unit/with_dbs/test_projects.py

```diff
@@ -1420,3 +1434,68 @@ async def test_open_shared_project_2_users_locked(
         expected.ok if user_role != UserRole.GUEST else web.HTTPOk,
         expected_project_state,
     )
+
+
+@pytest.mark.parametrize(*standard_role_response())
+async def test_open_shared_project_at_same_time(
+    loop,
+    client,
+    logged_user: Dict,
+    shared_project: Dict,
+    socketio_client: Callable,
+    client_session_id: Callable,
+    user_role: UserRole,
+    expected: ExpectedResponse,
+    aiohttp_client,
+):
+    # log client 1
+    client_1 = client
+    client_id1 = client_session_id()
+    sio_1 = await _connect_websocket(
+        socketio_client, user_role != UserRole.ANONYMOUS, client_1, client_id1,
+    )
+
+    # log client 2
+    client_2 = await aiohttp_client(client.app)
+    user_2 = await log_client_in(
+        client_2, {"role": user_role.name}, enable_check=user_role != UserRole.ANONYMOUS
+    )
+    client_id2 = client_session_id()
+    sio_2 = await _connect_websocket(
+        socketio_client, user_role != UserRole.ANONYMOUS, client_2, client_id2,
+    )
+
+    # try opening projects at same time (more or less)
+    results = await asyncio.gather(
+        _open_project(
+            client_1,
+            client_id1,
+            shared_project,
+            [
+                expected.ok if user_role != UserRole.GUEST else web.HTTPOk,
+                expected.locked if user_role != UserRole.GUEST else HTTPLocked,
+            ],
+        ),
+        _open_project(
+            client_2,
+            client_id2,
+            shared_project,
+            [
+                expected.ok if user_role != UserRole.GUEST else web.HTTPOk,
+                expected.locked if user_role != UserRole.GUEST else HTTPLocked,
+            ],
+        ),
+        return_exceptions=True,
+    )
+
+    # one should be opened, the other locked
+    if user_role != UserRole.ANONYMOUS:
+        num_assertions = 0
+        for data, error in results:
+            assert data or error
+            if error:
+                num_assertions += 1
+            elif data:
+                assert data == shared_project
+
+        assert num_assertions == 1
```

Comment:
why not to assert explicitly every result?
```python
DATA,ERROR=0,1

assert results[0][DATA] == shared_project and not result[0][ERROR]
assert not results[1][DATA] and result[1][ERROR]
...

```
This way you know which combination failed and not only that "SOME" combination failed

Reply:
hmm it can be one of both. how do you know which one is the locked one?? I could sort... but yeah...